File: algorithms/215.Kth_Largest_Element_in_an_Array/Kth_Largest_Element_in_an_Array.cpp
<<<<<<< HEAD
class Solution {

 public:
  int partition(vector<int> & nums, int lo, int hi) {
    int i = lo, j = hi + 1;
    while(1) {
      while(nums[++i] < nums[lo]) {
        if(i == hi) break;
      }
      while(nums[--j] > nums[lo]) {
        if(j == lo) break;
      }
      if(i >= j) break;
      swap(nums[i], nums[j]);
    }
    swap(nums[lo], nums[j]);
    return j;
  }

  // O(N) with worst cast O(N*N)
  int findKthLargest(vector<int>& nums, int k) {
    random_shuffle(nums.begin(), nums.end());
    int lo = 0, hi = nums.size() - 1;
    k = nums.size() - k;
    while(lo < hi) {
      int j = partition(nums, lo, hi);
      if(j < k) {
        lo = j + 1;
      } else if(j > k) {
          hi = j - 1;
      } else {
        return nums[k];
      }
    }
    return nums[k];
  }
};
class Solution { 
public:
    int partition(vector<int>& nums, int left, int right) {
        int pivot = nums[left];
        int l = left + 1, r = right;
        while (l <= r) {
            if (nums[l] < pivot && nums[r] > pivot)
                swap(nums[l++], nums[r--]);
            if (nums[l] >= pivot) l++; 
            if (nums[r] <= pivot) r--;
        }
        swap(nums[left], nums[r]);
        return r;
    }

    int findKthLargest(vector<int>& nums, int k) {
        int left = 0, right = nums.size() - 1;
        while (true) {
            int pos = partition(nums, left, right);
            if (pos == k - 1) return nums[pos];
            if (pos > k - 1) right = pos - 1;
            else left = pos + 1;
        }
    }
=======
class Solution {

 public:
  int partition(vector<int> & nums, int lo, int hi) {
    int i = lo, j = hi + 1;
    while(1) {
      while(nums[++i] < nums[lo]) {
        if(i == hi) break;
      }
      while(nums[--j] > nums[lo]) {
        if(j == lo) break;
      }
      if(i >= j) break;
      swap(nums[i], nums[j]);
    }
    swap(nums[lo], nums[j]);
    return j;
  }

  // O(N) with worst cast O(N*N)
  int findKthLargest(vector<int>& nums, int k) {
    random_shuffle(nums.begin(), nums.end());
    int lo = 0, hi = nums.size() - 1;
    k = nums.size() - k;
    while(lo < hi) {
      int j = partition(nums, lo, hi);
      if(j < k) {
        lo = j + 1;
      } else if(j > k) {
          hi = j - 1;
      } else {
        return nums[k];
      }
    }
    return nums[k];
  }
};
class Solution { 
public:
    int partition(vector<int>& nums, int left, int right) {
        int pivot = nums[left];
        int l = left + 1, r = right;
        while (l <= r) {
            if (nums[l] < pivot && nums[r] > pivot)
                swap(nums[l++], nums[r--]);
            if (nums[l] >= pivot) l++; 
            if (nums[r] <= pivot) r--;
        }
        swap(nums[left], nums[r]);
        return r;
    }

    int findKthLargest(vector<int>& nums, int k) {
        int left = 0, right = nums.size() - 1;
        while (true) {
            int pos = partition(nums, left, right);
            if (pos == k - 1) return nums[pos];
            if (pos > k - 1) right = pos - 1;
            else left = pos + 1;
        }
    }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
};

File: algorithms/215.Kth_Largest_Element_in_an_Array/Kth_Largest_Element_in_an_Array.java
public class KthLargestElement {
    public int findKthLargest(int[] nums, int k) {
        return findKthLargest(nums, k, 0, nums.length);
    }

    private int findKthLargest(int[] nums, int k, int start, int end) {
    	if (end - start < 2) return nums[start];
    	int mid = nums[end - 1];
    	int index = start - 1;
    	for (int i = start; i < end - 1; i++) {
    		if (nums[i] >= mid) {
    			exchange(nums, i, index + 1);
    			index++;
    		}
    	}
    	exchange(nums, index + 1, end - 1);
    	if (index + 1 == k - 1) {
    		return mid;
    	} else if (index + 1 < k - 1) {
    		 return findKthLargest(nums, k, index + 2, end);
    	} else {
    		return findKthLargest(nums, k, start, index + 1);
    	}
    }

    private void exchange(int[] nums, int i, int j) {
    	int temp = nums[i];
    	nums[i] = nums[j];
    	nums[j] = temp;
    }
}
public class Solution {

  public int findKthLargest(int[] a, int k) {
    int n = a.length;
    int p = quickSelect(a, 0, n - 1, n - k + 1);
    return a[p];
  }

  // return the index of the kth smallest number
  int quickSelect(int[] a, int lo, int hi, int k) {
    // use quick sort's idea
    // put nums that are <= pivot to the left
    // put nums that are  > pivot to the right
    int i = lo, j = hi, pivot = a[hi];
    while (i < j) {
      if (a[i++] > pivot) swap(a, --i, --j);
    }
    swap(a, i, hi);

    // count the nums that are <= pivot from lo
    int m = i - lo + 1;

    // pivot is the one!
    if (m == k)     return i;
    // pivot is too big, so it must be on the left
    else if (m > k) return quickSelect(a, lo, i - 1, k);
    // pivot is too small, so it must be on the right
    else            return quickSelect(a, i + 1, hi, k - m);
  }

  void swap(int[] a, int i, int j) {
    int tmp = a[i];
    a[i] = a[j];
    a[j] = tmp;
  }

}